3p/iframe-transport-client.js

```diff
@@ -33,61 +33,134 @@ export class IframeTransportClient {
+    this.iframeMessagingClient_.setSentinel(dev().assertString(
+        parsedFrameName['sentinel'],
         'Invalid/missing sentinel on iframe name attribute' + this.win_.name));
-    this.client_.makeRequest(
+    this.iframeMessagingClient_.makeRequest(
         MessageType.SEND_IFRAME_TRANSPORT_EVENTS,
         MessageType.IFRAME_TRANSPORT_EVENTS,
         eventData => {
           const events =
-              /**
-               * @type
-               * {!Array<../src/3p-frame-messaging.IframeTransportEvent>}
-               */
-              (eventData['events']);
-          user().assert(events,
+          /**
+           * @type
+           *   {!Array<../src/3p-frame-messaging.IframeTransportEvent>}
+           */
+          (eventData['events']);
+          dev().assert(events,
               'Received malformed events list in ' + this.win_.location.href);
           dev().assert(events.length,
               'Received empty events list in ' + this.win_.location.href);
-          user().assert(this.listener_,
-              'Must call onAnalyticsEvent in ' + this.win_.location.href);
           events.forEach(event => {
             try {
-              this.listener_ &&
-                  this.listener_(event.message, event.transportId);
+              dev().assert(event.creativeId,
+                  'Received malformed event in ' + this.win_.location.href);
+              this.contextFor_(event.creativeId).dispatch(event.message);
             } catch (e) {
               user().error(TAG_,
                   'Exception in callback passed to onAnalyticsEvent: ' +
-                  e.message);
+                  e);
             }
           });
         });
   }
 
   /**
+   * Retrieves/creates a context object to pass events pertaining to a
+   * particular creative.
+   * @param {string} creativeId The ID of the creative
+   * @returns {!IframeTransportContext}
+   * @private
+   */
+  contextFor_(creativeId) {
+    return this.creativeIdToContext_[creativeId] ||
+        (this.creativeIdToContext_[creativeId] =
+            new IframeTransportContext(this.win_, this.iframeMessagingClient_,
+            creativeId, this.vendor_));
+  }
+
+  /**
+   * Gets the IframeMessagingClient.
+   * @returns {!IframeMessagingClient}
+   * @VisibleForTesting
+   */
+  getIframeMessagingClient() {
+    return this.iframeMessagingClient_;
+  }
+}
+
+/**
+ * A context object to be passed along with event data.
+ */
+class IframeTransportContext {
+  /**
+   * @param {!Window} win
+   * @param {!IframeMessagingClient} iframeMessagingClient
+   * @param {string} creativeId The ID of the creative that the event
+   *     pertains to.
+   * @param {string} vendor The 3p vendor name
+   */
+  constructor(win, iframeMessagingClient, creativeId, vendor) {
+    /** @private {!IframeMessagingClient} */
+    this.iframeMessagingClient_ = iframeMessagingClient;
+
+    /** @private @const {!Object} */
+    this.baseMessage_ = {creativeId, vendor};
```

Comment:
We are not going to trust the vendor value here anyway, is it necessary?

Reply:
It is necessary, because now amp-ad-exit is listening for this response, and it needs to know the vendor name. It is used here:
https://github.com/ampproject/amphtml/pull/11305/files#diff-63cf316e6647dc9b56b2d7969d48f528R283
https://github.com/ampproject/amphtml/pull/11305/files#diff-63cf316e6647dc9b56b2d7969d48f528R152

Reply:
I suggest we store `this.creativeId_` and `this.vendor_` separately. And later when we compose a response object we can simply use 
```
dict({
  'creativeId': this.creativeId_,
  'vendor': this.vendor_,
  'data': data,
})
``` 
In this way, people know what the message object looks like directly, without having to study what a `baseMessage_` looks like. Up to you : )

Reply:
I had it like this originally, but yesterday @keithwrightbos suggested to change it to use baseMessage_. (https://github.com/ampproject/amphtml/pull/11306#discussion_r140840375)
Since the two of you disagree, I think I will leave it as-is.